src/lib.rs

```diff
@@ -49,557 +124,1353 @@ pub fn parse_document(buffer: &str) -> Node {
-        let manager = ResourceManager::make_shared();
-        let resource = manager.borrow_mut().resource_for(root_ptr);
-        let capabilities = CapabilityFactory::new()
-            .with_all()
-            .build(&resource, manager.clone());
+        let manager = Rc::new(ResourceManager::new());
+        manager.track_root(&root_ptr);
 
-        Node::new(capabilities)
+        Node::Document(Document {
+            resource: Resource {
+                pointer: root_ptr,
+                manager,
+            }
+        })
     }
 }
 
-/// Represents a node in the document tree.
-pub struct Node {
-    /// The set of capabilities this node possesses.
-    pub capabilities: NodeCapabilities,
+/// A node in the AST of a parsed commonmark document
+pub enum Node {
+    Document(Document),
+    BlockQuote(BlockQuote),
+    List(List),
+    Item(Item),
+    CodeBlock(CodeBlock),
+    HtmlBlock(HtmlBlock),
+    CustomBlock(CustomBlock),
+    Paragraph(Paragraph),
+    Heading(Heading),
+    ThematicBreak(ThematicBreak),
+    Text(Text),
+    SoftBreak(SoftBreak),
+    LineBreak(LineBreak),
+    Code(Code),
+    HtmlInline(HtmlInline),
+    CustomInline(CustomeInline),
+    Emph(Emph),
+    Strong(Strong),
+    Link(Link),
+    Image(Image)
 }
 
-impl Node {
-    /// Create a new Node instance from a set of capabilities.
-    fn new(capabilities: NodeCapabilities) -> Node {
-        Node { capabilities }
+impl PartialEq for Node {
+    fn eq(&self, other: &Node) -> bool {
+        self.as_resource().pointer == other.as_resource().pointer
     }
 }
 
-pub struct NodeFactory {
-    capability_factory: CapabilityFactory,
+impl Debug for Node {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        write!(f, "{} id: {:?}", self.get_cmark_type_string().unwrap_or("Type Unavailable".to_string()), self.as_resource().pointer)
+    }
 }
 
-/// The set of capabilities a Node instance possesses.
-///
-/// All methods available on a Node instance are invoked through the fields of this object.
-pub struct NodeCapabilities {
-    pub get: Option<NodeGetter>,
-    pub set: Option<NodeSetter>,
-    pub traverse: Option<NodeTraverser>,
-    pub mutate: Option<StructuralMutator>,
-    pub render: Option<NodeRenderer>,
-    destruct: Option<NodeDestructor>,
+impl Node {
+    /// Construct a Rust Node wrapper around a libcmark node
+    ///
+    /// The manager parameter should be a clone of the manager all other nodes are sharing.
+    fn new(pointer: *mut CMarkNodePtr, manager: Rc<ResourceManager>) -> DoogieResult<Self> {
+        let resource = Resource {
+            pointer,
+            manager
+        };
+
+        let cmark_type: NodeType;
+        unsafe {
+            cmark_type = NodeType::try_from(cmark_node_get_type(pointer) as u32)?;
+        }
+
+        match cmark_type {
+            NodeType::CMarkNodeNone => Err(DoogieError::NodeNone),
+            NodeType::CMarkNodeDocument => Ok(Node::Document(Document {resource})),
```

Comment:
What if you made a trait to help this?
```rust
trait NodeData { 
  fn wrap(self) -> Node
}
```

Or maybe you could do something with `std::convert::from`?  That should allow you to do something like
`NodeType::CMarkNodeDocument => Ok(Dcoument::new(resource).into())`